File: leetcode/algo2/day10/39_combination_sum.cpp
#include <vector>
#include <iostream>

void search(std::vector<int> &candidates, int pointer, int &target, std::vector<int> temp, int temp_sum, std::vector<std::vector<int>> &res)
{
    if (temp_sum == target)
    {
        res.push_back(temp);
        return;
    }
    if (pointer >= candidates.size() || temp_sum > target)
        return;
    temp.push_back(candidates[pointer]);
    search(candidates, pointer, target, temp, temp_sum + candidates[pointer], res);
    temp.pop_back();
    search(candidates, pointer + 1, target, temp, temp_sum, res);
}

std::vector<std::vector<int>> combinationSum(std::vector<int> &candidates, int target)
{
    int pointer = 0;
    std::vector<int> temp = {};
    int temp_sum = 0;
    std::vector<std::vector<int>> res;
    search(candidates, pointer, target, temp, temp_sum, res);
    return res;
}

int main()
{
    std::vector<int> nums = {2, 3, 6, 7};
    int target = 6;
    for (auto x : combinationSum(nums, target))
    {
        for (auto y : x)
            std::cout << y << " ";
        std::cout << std::endl;
    }
}

File: leetcode/algo2/day10/39_combination_sum.js
let res = [];
var solve = (candidates, target, temp_arr, pointer, sum) => {
    if (sum === target) {
        res.push([...temp_arr]);
        return;
    }
    if (pointer >= candidates.length || sum >= target) {
        return;
    }
    temp_arr.push(candidates[pointer]);
    solve(candidates, target, temp_arr, pointer, sum + candidates[pointer]);
    temp_arr.pop();
    solve(candidates, target, temp_arr, pointer + 1, sum);
}


var combinationSum = function (candidates, target) {
    solve(candidates, target, [], 0, 0);
    return res;
};


let candidates = [2, 3, 6, 7];
let target = 6;
for (const x of combinationSum(candidates, target)) {
    console.log(x);
}